tests/ext/drv/test_ext.c

```diff
@@ -0,0 +1,489 @@
+        status = STATUS_NO_MEMORY;
+        goto Exit;
+    }
+
+    hook_client->nmr_binding_handle = nmr_binding_handle;
+    hook_client->client_module_id = client_registration_instance->ModuleId->Guid;
+    hook_client->client_binding_context = client_binding_context;
+    hook_client->client_data = client_registration_instance->NpiSpecificCharacteristics;
+    client_dispatch_table = (ebpf_extension_dispatch_table_t*)client_dispatch;
+    if (client_dispatch_table == NULL) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+    hook_client->invoke_program = (ebpf_invoke_program_function_t)client_dispatch_table->function[0];
+
+    local_provider_context->attached_client = hook_client;
+
+Exit:
+
+    if (NT_SUCCESS(status)) {
+        *provider_binding_context = hook_client;
+        hook_client = NULL;
+    } else
+        ebpf_free(hook_client);
+
+    return status;
+}
+
+/**
+ * @brief Callback invoked when a hook NPI client detaches.
+ *
+ * @param client_binding_context Provider module's context for binding with the client.
+ * @retval STATUS_SUCCESS The operation succeeded.
+ * @retval STATUS_INVALID_PARAMETER One or more parameters are invalid.
+ */
+NTSTATUS
+_test_ebpf_extension_hook_provider_detach_client(void* provider_binding_context)
+{
+    NTSTATUS status = STATUS_SUCCESS;
+
+    test_ebpf_extension_hook_client_t* local_client_context =
+        (test_ebpf_extension_hook_client_t*)provider_binding_context;
+    test_ebpf_extension_hook_provider_t* provider_context = NULL;
+
+    if (local_client_context == NULL) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    provider_context = &_test_ebpf_extension_hook_provider_context;
+    provider_context->attached_client = NULL;
+
+    ebpf_free(local_client_context);
+
+Exit:
+    return status;
+}
+
+void
+test_ebpf_extension_hook_provider_unregister()
+{
+    test_ebpf_extension_hook_provider_t* provider_context = &_test_ebpf_extension_hook_provider_context;
+
+    NTSTATUS status = NmrDeregisterProvider(provider_context->nmr_provider_handle);
+    if (status == STATUS_PENDING)
+        // Wait for clients to detach.
+        NmrWaitForProviderDeregisterComplete(provider_context->nmr_provider_handle);
+}
+
+NTSTATUS
+test_ebpf_extension_hook_provider_register()
+{
+    test_ebpf_extension_hook_provider_t* local_provider_context;
+    NTSTATUS status = STATUS_SUCCESS;
+
+    local_provider_context = &_test_ebpf_extension_hook_provider_context;
+
+    status = NmrRegisterProvider(
+        &_test_ebpf_extension_hook_provider_characteristics,
+        local_provider_context,
+        &local_provider_context->nmr_provider_handle);
+    if (!NT_SUCCESS(status))
+        goto Exit;
+
+Exit:
+    if (!NT_SUCCESS(status))
+        test_ebpf_extension_hook_provider_unregister();
+
+    return status;
+}
+
+ebpf_result_t
+test_ebpf_extension_invoke_program(_In_ test_program_context_t* context, _Out_ uint32_t* result)
```

Comment:
should context be const?  Or should `_In_` be `_Inout_`?

Reply:
const. Fixed.